File: src/main/java/com/raychenon/leetcode/tree/InvertBinaryTree.java
package com.raychenon.leetcode.tree;

import java.util.LinkedList;
import java.util.Queue;

/**
 * User: raychenon
 * Date: 2019-10-21
 * https://leetcode.com/problems/invert-binary-tree/
 */
public class InvertBinaryTree {

    /**
     * swap the left and right child of all nodes in the tree.
     *
     * @param root
     * @return
     */
    public TreeNode invertTree(TreeNode root) {
        if (root == null) return null;

        TreeNode right = invertTree(root.right);
        TreeNode left = invertTree(root.left);

        root.left = right;
        root.right = left;

        return root;
    }

    /**
     * Swap the left and right child of all nodes in the tree
     *
     * time complexity: O(n)
     * space complexity: O(n
     * @param root
     * @return
     */
    public TreeNode invertTreeIter(TreeNode root) {
        if (root == null) return null;
        Queue<TreeNode> queue = new LinkedList<TreeNode>();
        queue.add(root);
        while (!queue.isEmpty()) {
            TreeNode current = queue.poll();
            TreeNode temp = current.left;
            current.left = current.right;
            current.right = temp;
            if (current.left != null) queue.add(current.left);
            if (current.right != null) queue.add(current.right);
        }

        return root;
    }
}

File: src/main/kotlin/com/raychenon/kotlin/leetcode/tree/InvertBinaryTree.kt
package com.raychenon.kotlin.leetcode.tree

import com.raychenon.leetcode.tree.TreeNode
import java.util.*

/**
 * User: raychenon
 * https://leetcode.com/problems/invert-binary-tree/
 */
object InvertBinaryTree {

    /**
     *  swap the left and right child of all nodes in the tree.
     *  Recursive.
     *  time complexity: O(n)
     *  space complexity: O(n)
     */
    fun invertTreeRecursive(root: TreeNode?): TreeNode? {
        if (root == null) return null

        var left = invertTreeRecursive(root?.left)
        var right = invertTreeRecursive(root?.right)

        root.left = right
        root.right = left

        return root
    }

    /**
     *  time complexity: O(n)
     *  space complexity: O(n)
     */
    fun invertTreeIter(root: TreeNode?): TreeNode? {
        if (root == null) return null

        var queue: Queue<TreeNode> = LinkedList<TreeNode>()
        queue.add(root)
        while (queue.isNotEmpty()) {

            var current = queue.poll()
            var temp = current?.left
            current.left = current?.right
            current.right = temp

            // current?.left.let { queue.add(it) }
            // current?.right.let { queue.add(it) }
            if (current.left != null) queue.add(current.left)
            if (current.right != null) queue.add(current.right)
        }
        return root
    }

}